### PR TITLE
Fix state concurrency issues

### PR DESCRIPTION
**Pull request by mihaitodor**
`catalog.ServiceState` does not tolerate concurrent read access

### CODE REVIEW

**Comment by bparli** [comment]
@mihaitodor Did you consider calling your mutex function from inside the respective state functions (state.ByService, state.Encode, state.EachService, etc)? It might be a little more of a safeguard that way.

**Comment by mihaitodor** [comment]
@bparli I have, but I hesitated because that might introduce some performance penalties in unexpected places. It might be better to consider refactoring the whole design of ServiceState, but not sure if it's worth the time investment.

**Comment by relistan** [comment]
I'm normally a supporter of decorators but with the use patterns for the state lock where we need to sometimes defer the unlock and sometimes not, I don't think it's a great fit here. The end result is some places would use the decorator while many wouldn't. Which just introduces inconsistency.

Regarding @bparli's point: actually we really don't want to put them inside the state functions. The reason is that there are places where the locking behavior is such that you need to call one of these functions from inside a function where you already hold the lock.

In the end the reason for this messiness is that I should have designed the state differently. I haven't found the time to redo the core of it around message passing instead of direct access and mutexes.

@mihaitodor let's back out the `SafeAccess` decorator and just put locks in the few places we need them.

**Comment by mihaitodor** [comment]
@relistan That's cool. I agree that using a decorator will be inconsistent, so I tossed it out :) Please have a look again at the changes.

**Comment by mihaitodor** [comment]
That's odd. This is the second time the automated build hanged in the same place, just before running the HAProxy tests. Any clue why this might happen? I can't reproduce it locally.

**Comment by relistan** [comment]
I also cannot and haven't been able to reproduce locally. Has been
happening randomly for awhile. Must be a race somewhere.

**Comment by mihaitodor** [comment]
Hopefully, all is fixed now :) @bparli please let me know if we're good to merge.

File: http.go [hunk 1]
```diff
@@ -88,6 +89,10 @@ func watchHandler(response http.ResponseWriter, req *http.Request, list *memberl
 
 	var jsonBytes []byte
 	pushUpdate := func() error {
+		// Enter critical section
+		state.RLock()
+		defer state.RUnlock()
```
Comment (bparli) [inline http.go]: Are you sure you want to hold the lock this long @mihaitodor?  It seems to me you could get away with releasing it after the if/else block.

Comment (mihaitodor) [inline http.go]: Indeed @bparli, it's probably not such a good idea to lock the state for so long here.

**Comment by mihaitodor** [comment]
@relistan Are we good to go with this?

**Pull request merged**